File: backend/app/schemas/analytics.py
"""Pydantic schemas for the three analytics features:
  - Momentum Swings
  - Clutch Index
  - Coach Decision Grades
"""
from __future__ import annotations

import uuid
from typing import Literal

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Shared
# ---------------------------------------------------------------------------

class PlayRef(BaseModel):
    """Minimal play reference used inside analytics objects."""
    play_id: uuid.UUID
    sequence: int
    quarter: int
    game_clock_seconds: int
    description: str | None


# ---------------------------------------------------------------------------
# 1. Momentum Swings
# ---------------------------------------------------------------------------

class MomentumSwing(BaseModel):
    rank: int
    play_ref: PlayRef
    wp_before: float
    wp_after: float
    delta_wp: float          # signed: positive = helped home
    magnitude: float         # abs(delta_wp)
    tag: str | None          # "turnover" | "touchdown" | "field_goal" | "fourth_down" | None
    is_turning_point: bool   # True only for rank == 1


class MomentumSwingsResponse(BaseModel):
    game_id: uuid.UUID
    swings: list[MomentumSwing]


# ---------------------------------------------------------------------------
# 2. Clutch Index
# ---------------------------------------------------------------------------

class ClutchPlay(BaseModel):
    rank: int
    play_ref: PlayRef
    delta_wp: float
    clutch_score: float
    time_factor: float
    close_factor: float
    score_diff: int           # home - away at the time of the play


class ClutchDrive(BaseModel):
    drive_number: int         # sequential within game (derived from possession flips)
    posteam_abbr: str | None
    clutch_total: float
    play_count: int


class ClutchTeamTotals(BaseModel):
    offense: float
    defense: float


class ClutchResponse(BaseModel):
    game_id: uuid.UUID
    top_plays: list[ClutchPlay]
    top_drives: list[ClutchDrive]
    team_totals: dict[str, ClutchTeamTotals]  # keys: "home" | "away"


# ---------------------------------------------------------------------------
# 3. Coach Decision Grades
# ---------------------------------------------------------------------------

class DecisionOption(BaseModel):
    wp: float
    detail: str | None = None      # e.g. "p_conv=0.63" or "expected_net=41 yds"


class CoachDecision(BaseModel):
    play_ref: PlayRef
    situation: str                  # "4th & 2 at OPP 43"
    actual_type: Literal["go_for_it", "punt", "field_goal"]
    actual_wp_after: float
    alternatives: dict[str, DecisionOption | None]  # key = action name
    best_action: str
    decision_delta: float           # actual_wp - best_wp  (≤ 0)
    grade: Literal["Optimal", "Questionable", "Bad", "Very Bad"]
    grade_emoji: str


class DecisionGradesResponse(BaseModel):
    game_id: uuid.UUID
    decisions: list[CoachDecision]

File: frontend/src/types/analytics.ts
// Analytics types matching the backend app/schemas/analytics.py models

export interface PlayRef {
  play_id: string
  sequence: number
  quarter: number
  game_clock_seconds: number
  description: string | null
}

// -----------------------------------------------------------------------
// Momentum Swings
// -----------------------------------------------------------------------

export interface MomentumSwing {
  rank: number
  play_ref: PlayRef
  wp_before: number
  wp_after: number
  delta_wp: number
  magnitude: number
  tag: string | null
  is_turning_point: boolean
}

export interface MomentumSwingsResponse {
  game_id: string
  swings: MomentumSwing[]
}

// -----------------------------------------------------------------------
// Clutch Index
// -----------------------------------------------------------------------

export interface ClutchPlay {
  rank: number
  play_ref: PlayRef
  delta_wp: number
  clutch_score: number
  time_factor: number
  close_factor: number
  score_diff: number
}

export interface ClutchDrive {
  drive_number: number
  posteam_abbr: string | null
  clutch_total: number
  play_count: number
}

export interface ClutchTeamTotals {
  offense: number
  defense: number
}

export interface ClutchResponse {
  game_id: string
  top_plays: ClutchPlay[]
  top_drives: ClutchDrive[]
  team_totals: {
    home: ClutchTeamTotals
    away: ClutchTeamTotals
  }
}

// -----------------------------------------------------------------------
// Coach Decision Grades
// -----------------------------------------------------------------------

export interface DecisionOption {
  wp: number
  detail: string | null
}

export type DecisionType = 'go_for_it' | 'punt' | 'field_goal'
export type GradeLabel = 'Optimal' | 'Questionable' | 'Bad' | 'Very Bad'

export interface CoachDecision {
  play_ref: PlayRef
  situation: string
  actual_type: DecisionType
  actual_wp_after: number
  alternatives: Record<string, DecisionOption | null>
  best_action: string
  decision_delta: number
  grade: GradeLabel
  grade_emoji: string
}

export interface DecisionGradesResponse {
  game_id: string
  decisions: CoachDecision[]
}
